### PR TITLE
Increase table serialization limits

### LINKED ISSUE #6676: Add an option to increase limits of `sys.serialize` and `sys.save`
**Is your feature request related to a problem? Please describe (REQUIRED):**
Sometimes it's necessary to serialize or to save large Lua tables, but, currently, `sys.serialize` and `sys.save` functions are limited by 512KB of memory. Plus, the total number of rows that any one table may contain is limited to 65536 (i.e. a 16 bit range).

**Describe the solution you'd like (REQUIRED):**
Add an option as the second/third argument to increase limits for a single function call:
```lua
-- sys.serialize(table, [options])
-- sys.save(filename, table, [options])

local result = sys.serialize(super_large_table, { allocate_memory = 10 * 1024 * 1024, range_bits = 32 })
```

P.S. It's pretty simple to implement an option to set memory limit, i.e. malloc/free temporary buffer like that https://github.com/aglitchman/defold-savetable/blob/main/savetable/src/main.cpp#L22

### PR DESCRIPTION
**Pull request by britzl**
It is now possible to serialize Lua tables with more than 65535 keys (new max is 2^32). The previous maximum size of 512kb for the serialized table has been removed and dynamic memory allocation will be used for tables larger than 512kb.

Fixes #6676

### CODE REVIEW

**Comment by britzl** [comment]
Step 1 was to increase the key count from 16 to 32 bit. Next step is to change how the buffer save and load buffer is created to allow serialization of larger tables. I'm thinking that it might make sense to dynamically allocate the buffer instead of hardcoding it.

**Comment by britzl** [comment]
> Next step is to change how the buffer save and load buffer is created to allow serialization of larger tables. I'm thinking that it might make sense to dynamically allocate the buffer instead of hardcoding it.

I've added a function to calculate the size of the table and then dynamically allocate the memory needed if above 512kb

File: engine/script/src/test/test_script_table.cpp [hunk 1]
```diff
@@ -168,7 +168,7 @@ TEST_F(LuaTableTest, AttemptReadUnsupportedVersion)
     int result = lua_cpcall(L, ReadUnsupportedVersion, 0x0);
     ASSERT_NE(0, result);
     char str[256];
-    dmSnPrintf(str, sizeof(str), "Unsupported serialized table data: version = 0x%x (current = 0x%x)", 818192, 3);
+    dmSnPrintf(str, sizeof(str), "Unsupported serialized table data: version = 0x%x (current = 0x%x)", 818192, 4);
```
Comment (AGulev) [inline engine/script/src/test/test_script_table.cpp:171]: Does that means we can't read old saves anymore? 
If we can, do we have a test that check this?

Comment (britzl) [reply engine/script/src/test/test_script_table.cpp:171]: No, no, we can still read old versions. BUT I don't think I have a test for reading a v3 table with the v4 serializer. I might as well add one!

File: build_tools/run.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def _exec_command(arg_list, **kwargs):
 
         output = ''
         while True:
-            line = process.stdout.readline().decode()
+            line = process.stdout.readline().decode(errors='replace')
```
Comment (britzl) [inline build_tools/run.py:47]: I got an exception here with utf-8 vs latin-1 encoding on Windows when logging some text. T

File: engine/script/src/script_sys.cpp [hunk 3]
```diff
@@ -70,6 +71,26 @@ union SaveLoadBuffer
      * @namespace sys
      */
 
+    char* Sys_SetupTableSerializationBuffer(int required_size)
```
Comment (britzl) [inline engine/script/src/script_sys.cpp:74]: Use the 512kb fixed buffer for "small" files and a dynamically allocated buffer for large files

File: engine/script/src/test/test_script_sys.cpp [hunk 4]
```diff
@@ -92,7 +92,6 @@ bool RunString(lua_State* L, const char* script)
 }
 
 
-#if defined(__NX__) && !defined(GITHUB_CI) // we cannot run this test unless we have a device connected
```
Comment (britzl) [inline engine/script/src/test/test_script_sys.cpp:95]: @JCash I'm a bit confused by this. Why don't we want to run these tests for all platforms?

Comment (JCash) [reply engine/script/src/test/test_script_sys.cpp:95]: Not sure.
Possibly something I had to disable at the time

**Review by JCash — DISMISSED** [review]

**Pull request merged**